### PR TITLE
Localisation and reworking currency

### PR DESCRIPTION
**Pull request by sfount**
Attempt 2.

In this PR:

1. Localisation using ngLocale, configures angular filters based on locale - the primary change has been updating dates from only US format to respecting the user language. __(Note: this does not apply to the currency filter - as described in this pull request)__
2. Forked the native angular currency filter - it is essential that currencies in BHIMA are __not__ dependent on locale, they must respect the original transactions currency. This filter enables:
  * Safe alternative to intlcurrency, no html binding is used
  * Filter independently of current locale
  * Specifiy BHIMA currency ID to request formatting
  * Render strong warnings if a non-supported currency is used or a mistake is made 
  * Asynchronously filter based on available JSON configuration files
3. Updated the database to remove formatting from currency line and create relationship with format configuration files
4. Service to manage async fetching and requesting of currency format files. This can be used by any module that requires information about currency format information (selects showing available currency options) however is primarily intended to drive the currency filter. 
5. Deprecated the intlcurrency filter - please note this PR request does not replace all instances of the filter. For now it simple returns an error message on use. 

An example of how the currency filter can be used
```
{{ transaction.total | currency:transaction.currency }}
```

The currency filter provides the configuration boolean ```requiresCurrencyDefinition``` By default this is set to false and the enterprise currency is used if no currency is specified. If this flag is set the filter will display a clear error stating that the developer should explicitly define a currency id. The development team can decide if it is safe to allow implicit use of the enterprise currency.

If this change is accepted I am happy to merge the very small conflicts between this branch and ```development```. (See my attempt at being clever here #688).

### CODE REVIEW

**Comment by sfount** [comment]
Addresses #569

File: client/src/js/app.js [hunk 1]
```diff
@@ -746,11 +746,17 @@
...
-    //TODO Try and assign the previous sessions language key here
+  
     $translateProvider.preferredLanguage('fr');
   }
 
+  function localeConfig(tmhDynamicLocaleProvider) { 
+
+    // TODO Hardcoded default translation/ localisation
+    tmhDynamicLocaleProvider.localeLocationPattern('/i18n/locale/angular-locale_{{locale}}.js');
+    tmhDynamicLocaleProvider.defaultLocale('fr-fr');
```
Comment (jniles) [inline client/src/js/app.js]: This is minor, but it might be best to bundle the [Belgian French](https://github.com/angular/angular.js/blob/master/src/ngLocale/angular-locale_fr-be.js) locale as default, rather than the France French.

File: client/src/js/filters/filters.js [hunk 2]
```diff
@@ -1,109 +1,111 @@
...
+  .filter('boolean', function() {
+    return function (input) {
+      return Boolean(Number(input));
+    };
+  })
+
+  // Potential currency filter providing a specified locale 
+  // Ideally would leverage angular currency filter but this does NOT support
+  // passing through individual locale and doesn't expose utility methods 
+  .filter('intlcurrency', [
```
Comment (jniles) [inline client/src/js/filters/filters.js]: Is this filter still necessary?  In your [PR](https://github.com/IMA-WorldHealth/bhima/pull/689) you make no mention of it.  If not, can it be safely removed (all the modules have been migrated, or course)?

File: client/src/js/filters/filters.js [hunk 3]
```diff
@@ -1,109 +1,111 @@
...
-          if (decimalDigits) {
-            value = value.slice(0, value.indexOf('.'));
-          }
+        // var templ = value.replace(/(\d)(?=(\d\d\d)+(?!\d))/g, '$1'+currency.get(id).separator);
+        // templ += '<span class="desc">' + currency.get(id).decimal + decimalDigits + '</span><span class="cur"> ' + currency.get(id).symbol +  '</span>';
 
-          var templ = value.replace(/(\d)(?=(\d\d\d)+(?!\d))/g, '$1'+currency.get(id).separator);
-          templ += '<span class="desc">' + currency.get(id).decimal + decimalDigits + '</span><span class="cur"> ' + currency.get(id).symbol +  '</span>';
+        // return $sce.trustAsHtml(templ);
+        return $sce.trustAsHtml('<span style="color : red;">Filter Depriciated</span>');
```
Comment (jniles) [inline client/src/js/filters/filters.js]: Welp.  I should have read more code.  Ignore [the comment above](https://github.com/IMA-WorldHealth/bhima/pull/689/files#r38726482).

File: client/src/partials/receipts/receipt.sale.js [hunk 4]
```diff
@@ -1,13 +1,14 @@
 angular.module('bhima.controllers')
 .controller('receipt.sale', [
   '$scope',
+  '$locale',
```
Comment (jniles) [inline client/src/partials/receipts/receipt.sale.js]: Where is `$locale` used in the sale receipt?

**Comment by jniles** [comment]
Minor fixes, but a merge with development.  On the whole, it looks awesome!  Can't wait to use this!

File: client/src/js/app.js (hunk 1 above)
Comment (sfount) [inline client/src/js/app.js]: No problem with that. This does show a fairly lazy thing we do here using 'fr'/ 'fr-fr' strings as default.
This doesn't work if you don't have French for example. It's not the end of the world because it should just default back to angular's default (en-US) if it fails but the feature is clearly useful. 

My first naive thoughts are that you could template into the file/ index.html the enterprises default language and locale. This way it is ready to use during the angular bootstrap process. (app.js). (process mentioned brielfy here #663 )

File: client/src/js/filters/filters.js (hunk 2 above)
Comment (sfount) [inline client/src/js/filters/filters.js]: Although it was an essay:
'5. Deprecated the intlcurrency filter - please note this PR request does not replace all instances of the filter. For now it simple returns an error message on use.'

This was mostly just a time issue. I grepped 'intlcurrency' and decided I would make it a separate commit!

File: client/src/js/app.js (hunk 1 above)
Comment (jniles) [inline client/src/js/app.js]: True, true.  It was just an observation.

We should research a proper way to bootstrap the app.  The naive (probably easiest) way would be at build time with something like `gulp build --production --locales=en-US,fr-FR,fr-BE`.  I've heard some things about Angular2's router being about to lazy load routes, so maybe we can expect more support for bootstrapping dynamically within Angular.  Of course, we could try to bootstrap before Angular is even loaded, that idea is probably the worst in terms of clarity, maintainability, and portability.

File: client/src/js/filters/filters.js (hunk 2 above)
Comment (jniles) [inline client/src/js/filters/filters.js]: Yup. My bad.  Overzealously posted before thoroughly reading. :S

Comment (sfount) [inline client/src/js/filters/filters.js]: Hey no worries, I appreciate how quickly you got around to it and the time that this kind of review takes.
